### ISSUE TITLE
ResizeObserver (and MutationObserver) not executed while in fakeAsync

### ISSUE THREAD

**Issue by Goodwine** [angular/zone.js#1144]
## Current behavior
Example with delay (my use case) -
https://stackblitz.com/edit/angular-f8evbx?file=src/app/app.component.spec.ts
Simplified example -
https://stackblitz.com/edit/angular-f8evbx-3ak8vv?file=src/app/app.component.spec.ts
MutationObserver example -
https://stackblitz.com/edit/angular-f8evbx-6wdpmj?file=src/app/app.component.ts

1. Use ResizeObserver (or MutationObserver + window.onresize to polyfill) on an element
2. (not relevant) Handle event into an observable with debouce or some other delay thingy
3. Write test for it using fakeAsync

Tests using fakeAsync fail, I'm forced to use `setTimeout(expectations+done, delay)` which causes unexpected delays. I tried using `import 'zone.js/dist/zone-patch-resize-observer';` but it didn't work.

On top of that, the timeout is flaky unless a full extra second or so is added to the delay.

## Expected behavior

`flush()` should trigger events from `ResizeObserver` and `MutationObserver`

## Other things
"zone.js": "^0.8.26"
"@angular/core": "^6.1.0",
"rxjs": "^6.2.0",

Relevant: https://github.com/angular/zone.js/pull/1012

**Comment by Goodwine** [comment]
I also tried `whenStable` just in case but without any hope, it didn't work either

**Comment by JiaLiPassion** [comment]
Thanks for posting the issue, The reason `MutationObserver`, `ResizeObserver` not work in `fakeAsync` because they are not `patched` as `Task` like `setTimeout/Promise`. I will consider how to turn them into `task`.

**Comment by Goodwine** [comment]
Hiya, I was wondering if this is something I could try to work on, do you have an example commit showing how a patch like this is done?

**Comment by JiaLiPassion** [comment]
@Goodwine, thanks a lot, I am working on this now, it involves `MutiationObserver, ResizeObserver` and some other stuff such as `FileReader`, I want to make a generic routine to support them in `fakeAsync`, and thanks again, I will let you know after I finish.

**Comment by Goodwine** [comment]
I noticed this also being commented on https://github.com/angular/zone.js/issues/1011 which is an earlier bug marked as closed/fixed, is it actually fixed? or no? would the workaround mentioned on that other Issue work here?

**Comment by Akxe** [comment]
@Goodwine It is not, the issue is about `ResizeObserver` not calling Angular's change detection, but this issue is still not solved.
I am not sure how to actually test code that is dependant on `ResizeObserver`.

**Comment by Goodwine** [comment]
While this is not implemented, you should be able to grab the class and replace it, something like this:
```ts
let backup: typeof ResizeObserver;
let callback: () => void;
// take over ResizeObserver
beforeAll(() => {
  backup = ResizeObserver;
  (ResizeObserver as unknown) = class {
    constructor(c: () => void) {
      callback = c;
    }
    observe = jasmine.createSpy('observe');
    disconnect = jasmine.createSpy('disconnect');
  }
});
// restore ResizeObserver
afterAll(() => { (ResizeObserver as unknown) = backup });
// reset callback
beforeEach(() => {
  callback = () => {}
});
// test
it(() => {
  doSomething();
  expect(something).toBe(false);
  callback();
  expect(something).toBe(true);
});
```

**Comment by Akxe** [comment]
@Goodwine That's true, the only bad thing is, that I use `resize-observer-polyfill`, and by importing the component containing the polyfill, it gets set to `windows.ResizeObserver`, by reference. I now only need to find a way to replace it sooner than the tested code gets its grasp on it. (Will probably share the solution, as I can see how others would end up in the same situation...)

**Comment by Goodwine** [comment]
It should still work with polyfilled classes, this is meant to be used on the `*_test.ts` file and not at a global level. And it's grabbing the "ambient" value of `ResizeObserver` which should be the same as `window.ResizeObserver`. But otherwise you can change the code to grab the value from the window instead of from the ambient/global state.

That said, this is just my workaround while we wait for the real implementation in zonejs
